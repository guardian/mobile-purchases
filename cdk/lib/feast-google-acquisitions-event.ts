import type { GuStackProps } from "@guardian/cdk/lib/constructs/core";
import { GuStack } from "@guardian/cdk/lib/constructs/core";
import { GuSnsLambdaExperimental } from "@guardian/cdk/lib/experimental/patterns";
import { Duration } from 'aws-cdk-lib';
import type { App } from "aws-cdk-lib";
import { Runtime } from 'aws-cdk-lib/aws-lambda';
import { Queue } from "aws-cdk-lib/aws-sqs";

export interface FeastSubscriptionsProps extends GuStackProps {
    stack: string;
    stage: string;
}


export class FeastGoogleAcquisitionsEvent extends GuStack {
    constructor(scope: App, id: string, props: FeastSubscriptionsProps) {
        super(scope, id, props);

        const app = 'mobile-purchases-feast-google-acquisitions-event';
        const nameWithStage = `${app}-${this.stage}`;

        const feastGoogleAcquisitionsEvent = new Queue(this, `${nameWithStage}-sns-queue`, {
            queueName: `${nameWithStage}-queue`,
            retentionPeriod: Duration.days(14),
        });

        new GuSnsLambdaExperimental(this,`${ nameWithStage }`,{
            app: `${ nameWithStage }`,
            functionName: `${ nameWithStage }`,
            existingSnsTopic: { externalTopicName: feastGoogleAcquisitionsEvent.queueName },
            runtime: Runtime.NODEJS_20_X,
            monitoringConfiguration: { noMonitoring: true },
            handler: 'feast-google-acquisitions-event.handler',
            fileName: `feast-google-acquisitions-event.zip`,
        });

        //permissions to read sns queue

        //permissions to access SMS and get google auth

        //permissions to make call to Google

        //permissions to read and write to database


    }

}