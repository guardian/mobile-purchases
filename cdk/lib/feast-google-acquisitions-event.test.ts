import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import {FeastGoogleAcquisitionsEvent } from "./feast-google-acquisitions-event";

describe('The FeastAndroidSubscriptions stack', () => {
    it('matches the snapshot', () => {
        const app = new App();
        const codeStack = new FeastGoogleAcquisitionsEvent(app, 'feast-subscriptions-CODE',{ stack: 'mobile', stage: 'CODE' });
        const prodStack = new FeastGoogleAcquisitionsEvent(app, 'feast-subscriptions-PROD',{ stack: 'mobile', stage: 'PROD' });

        expect(Template.fromStack(codeStack).toJSON()).toMatchSnapshot();
        expect(Template.fromStack(prodStack).toJSON()).toMatchSnapshot();
    });
});