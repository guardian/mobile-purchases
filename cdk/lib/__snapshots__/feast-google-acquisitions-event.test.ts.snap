// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The FeastAndroidSubscriptions stack matches the snapshot 1`] = `
{
  "Metadata": {
    "gu:cdk:constructs": [
      "GuDistributionBucketParameter",
      "GuSnsLambdaExperimental",
    ],
    "gu:cdk:version": "TEST",
  },
  "Outputs": {
    "mobilepurchasesfeastgoogleacquisitionseventCODETopicName2E26AB35": {
      "Value": {
        "Fn::GetAtt": [
          "mobilepurchasesfeastgoogleacquisitionseventCODEsnsqueueBBB527C9",
          "QueueName",
        ],
      },
    },
  },
  "Parameters": {
    "DistributionBucketName": {
      "Default": "/account/services/artifact.bucket",
      "Description": "SSM parameter containing the S3 bucket name holding distribution artifacts",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "mobilepurchasesfeastgoogleacquisitionseventCODE0FD07B8A": {
      "DependsOn": [
        "mobilepurchasesfeastgoogleacquisitionseventCODEServiceRoleDefaultPolicy0CCA2866",
        "mobilepurchasesfeastgoogleacquisitionseventCODEServiceRoleCC12270B",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Ref": "DistributionBucketName",
          },
          "S3Key": "mobile/CODE/mobile-purchases-feast-google-acquisitions-event-CODE/mobile-purchases-feast-google-acquisitions-event-CODE.zip",
        },
        "Environment": {
          "Variables": {
            "APP": "mobile-purchases-feast-google-acquisitions-event-CODE",
            "STACK": "mobile",
            "STAGE": "CODE",
          },
        },
        "FunctionName": "mobile-purchases-feast-google-acquisitions-event-CODE",
        "Handler": "mobile-purchases-feast-google-acquisitions-event-CODE.handler",
        "MemorySize": 512,
        "Role": {
          "Fn::GetAtt": [
            "mobilepurchasesfeastgoogleacquisitionseventCODEServiceRoleCC12270B",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
        "Tags": [
          {
            "Key": "App",
            "Value": "mobile-purchases-feast-google-acquisitions-event-CODE",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/mobile-purchases",
          },
          {
            "Key": "Stack",
            "Value": "mobile",
          },
          {
            "Key": "Stage",
            "Value": "CODE",
          },
        ],
        "Timeout": 30,
      },
      "Type": "AWS::Lambda::Function",
    },
    "mobilepurchasesfeastgoogleacquisitionseventCODEAllowInvokefeastsubscriptionsCODEmobilepurchasesfeastgoogleacquisitionseventCODESnsExistingIncomingEventsTopic418B4E36C1BEE404": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "mobilepurchasesfeastgoogleacquisitionseventCODE0FD07B8A",
            "Arn",
          ],
        },
        "Principal": "sns.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:aws:sns:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Fn::GetAtt": [
                  "mobilepurchasesfeastgoogleacquisitionseventCODEsnsqueueBBB527C9",
                  "QueueName",
                ],
              },
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "mobilepurchasesfeastgoogleacquisitionseventCODEServiceRoleCC12270B": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Tags": [
          {
            "Key": "App",
            "Value": "mobile-purchases-feast-google-acquisitions-event-CODE",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/mobile-purchases",
          },
          {
            "Key": "Stack",
            "Value": "mobile",
          },
          {
            "Key": "Stage",
            "Value": "CODE",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "mobilepurchasesfeastgoogleacquisitionseventCODEServiceRoleDefaultPolicy0CCA2866": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                      "/mobile/CODE/mobile-purchases-feast-google-acquisitions-event-CODE/mobile-purchases-feast-google-acquisitions-event-CODE.zip",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "ssm:GetParametersByPath",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/CODE/mobile/mobile-purchases-feast-google-acquisitions-event-CODE",
                  ],
                ],
              },
            },
            {
              "Action": [
                "ssm:GetParameters",
                "ssm:GetParameter",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/CODE/mobile/mobile-purchases-feast-google-acquisitions-event-CODE/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "mobilepurchasesfeastgoogleacquisitionseventCODEServiceRoleDefaultPolicy0CCA2866",
        "Roles": [
          {
            "Ref": "mobilepurchasesfeastgoogleacquisitionseventCODEServiceRoleCC12270B",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "mobilepurchasesfeastgoogleacquisitionseventCODEmobilepurchasesfeastgoogleacquisitionseventCODESnsExistingIncomingEventsTopicF7D47457": {
      "Properties": {
        "Endpoint": {
          "Fn::GetAtt": [
            "mobilepurchasesfeastgoogleacquisitionseventCODE0FD07B8A",
            "Arn",
          ],
        },
        "Protocol": "lambda",
        "Region": {
          "Ref": "AWS::Region",
        },
        "TopicArn": {
          "Fn::Join": [
            "",
            [
              "arn:aws:sns:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Fn::GetAtt": [
                  "mobilepurchasesfeastgoogleacquisitionseventCODEsnsqueueBBB527C9",
                  "QueueName",
                ],
              },
            ],
          ],
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "mobilepurchasesfeastgoogleacquisitionseventCODEsnsqueueBBB527C9": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "MessageRetentionPeriod": 1209600,
        "QueueName": "mobile-purchases-feast-google-acquisitions-event-CODE-queue",
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/mobile-purchases",
          },
          {
            "Key": "Stack",
            "Value": "mobile",
          },
          {
            "Key": "Stage",
            "Value": "CODE",
          },
        ],
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
  },
}
`;

exports[`The FeastAndroidSubscriptions stack matches the snapshot 2`] = `
{
  "Metadata": {
    "gu:cdk:constructs": [
      "GuDistributionBucketParameter",
      "GuSnsLambdaExperimental",
    ],
    "gu:cdk:version": "TEST",
  },
  "Outputs": {
    "mobilepurchasesfeastgoogleacquisitionseventPRODTopicName68759315": {
      "Value": {
        "Fn::GetAtt": [
          "mobilepurchasesfeastgoogleacquisitionseventPRODsnsqueueE303EB89",
          "QueueName",
        ],
      },
    },
  },
  "Parameters": {
    "DistributionBucketName": {
      "Default": "/account/services/artifact.bucket",
      "Description": "SSM parameter containing the S3 bucket name holding distribution artifacts",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "mobilepurchasesfeastgoogleacquisitionseventPROD9454A390": {
      "DependsOn": [
        "mobilepurchasesfeastgoogleacquisitionseventPRODServiceRoleDefaultPolicy50312786",
        "mobilepurchasesfeastgoogleacquisitionseventPRODServiceRole255E9D72",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Ref": "DistributionBucketName",
          },
          "S3Key": "mobile/PROD/mobile-purchases-feast-google-acquisitions-event-PROD/mobile-purchases-feast-google-acquisitions-event-PROD.zip",
        },
        "Environment": {
          "Variables": {
            "APP": "mobile-purchases-feast-google-acquisitions-event-PROD",
            "STACK": "mobile",
            "STAGE": "PROD",
          },
        },
        "FunctionName": "mobile-purchases-feast-google-acquisitions-event-PROD",
        "Handler": "mobile-purchases-feast-google-acquisitions-event-PROD.handler",
        "MemorySize": 512,
        "Role": {
          "Fn::GetAtt": [
            "mobilepurchasesfeastgoogleacquisitionseventPRODServiceRole255E9D72",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
        "Tags": [
          {
            "Key": "App",
            "Value": "mobile-purchases-feast-google-acquisitions-event-PROD",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/mobile-purchases",
          },
          {
            "Key": "Stack",
            "Value": "mobile",
          },
          {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
        "Timeout": 30,
      },
      "Type": "AWS::Lambda::Function",
    },
    "mobilepurchasesfeastgoogleacquisitionseventPRODAllowInvokefeastsubscriptionsPRODmobilepurchasesfeastgoogleacquisitionseventPRODSnsExistingIncomingEventsTopicB79D802A749AB53A": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "mobilepurchasesfeastgoogleacquisitionseventPROD9454A390",
            "Arn",
          ],
        },
        "Principal": "sns.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:aws:sns:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Fn::GetAtt": [
                  "mobilepurchasesfeastgoogleacquisitionseventPRODsnsqueueE303EB89",
                  "QueueName",
                ],
              },
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "mobilepurchasesfeastgoogleacquisitionseventPRODServiceRole255E9D72": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Tags": [
          {
            "Key": "App",
            "Value": "mobile-purchases-feast-google-acquisitions-event-PROD",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/mobile-purchases",
          },
          {
            "Key": "Stack",
            "Value": "mobile",
          },
          {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "mobilepurchasesfeastgoogleacquisitionseventPRODServiceRoleDefaultPolicy50312786": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                      "/mobile/PROD/mobile-purchases-feast-google-acquisitions-event-PROD/mobile-purchases-feast-google-acquisitions-event-PROD.zip",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "ssm:GetParametersByPath",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/PROD/mobile/mobile-purchases-feast-google-acquisitions-event-PROD",
                  ],
                ],
              },
            },
            {
              "Action": [
                "ssm:GetParameters",
                "ssm:GetParameter",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/PROD/mobile/mobile-purchases-feast-google-acquisitions-event-PROD/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "mobilepurchasesfeastgoogleacquisitionseventPRODServiceRoleDefaultPolicy50312786",
        "Roles": [
          {
            "Ref": "mobilepurchasesfeastgoogleacquisitionseventPRODServiceRole255E9D72",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "mobilepurchasesfeastgoogleacquisitionseventPRODmobilepurchasesfeastgoogleacquisitionseventPRODSnsExistingIncomingEventsTopic283F60DE": {
      "Properties": {
        "Endpoint": {
          "Fn::GetAtt": [
            "mobilepurchasesfeastgoogleacquisitionseventPROD9454A390",
            "Arn",
          ],
        },
        "Protocol": "lambda",
        "Region": {
          "Ref": "AWS::Region",
        },
        "TopicArn": {
          "Fn::Join": [
            "",
            [
              "arn:aws:sns:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Fn::GetAtt": [
                  "mobilepurchasesfeastgoogleacquisitionseventPRODsnsqueueE303EB89",
                  "QueueName",
                ],
              },
            ],
          ],
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "mobilepurchasesfeastgoogleacquisitionseventPRODsnsqueueE303EB89": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "MessageRetentionPeriod": 1209600,
        "QueueName": "mobile-purchases-feast-google-acquisitions-event-PROD-queue",
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/mobile-purchases",
          },
          {
            "Key": "Stack",
            "Value": "mobile",
          },
          {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
  },
}
`;
